Implementing Decision Trees with Python Scikit Learn

We will be exceuting both Classification and Regression using Decision Tree

#1. Decision Tree for Classification

we will predict whether a bank note is authentic or fake depending upon the four different attributes of the image of the note.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
dataset = pd.read_csv("/content/bill_authentication.csv")

In [3]:
dataset.head()

Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1372 entries, 0 to 1371
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Variance  1372 non-null   float64
 1   Skewness  1372 non-null   float64
 2   Curtosis  1372 non-null   float64
 3   Entropy   1372 non-null   float64
 4   Class     1372 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 53.7 KB


In [5]:
dataset.describe()

Variance     Skewness     Curtosis      Entropy        Class
count  1372.000000  1372.000000  1372.000000  1372.000000  1372.000000
mean      0.433735     1.922353     1.397627    -1.191657     0.444606
std       2.842763     5.869047     4.310030     2.101013     0.497103
min      -7.042100   -13.773100    -5.286100    -8.548200     0.000000
25%      -1.773000    -1.708200    -1.574975    -2.413450     0.000000
50%       0.496180     2.319650     0.616630    -0.586650     0.000000
75%       2.821475     6.814625     3.179250     0.394810     1.000000
max       6.824800    12.951600    17.927400     2.449500     1.000000

In [6]:
dataset.shape

(1372, 5)

In [7]:
# Preparing the data 

X = dataset.drop('Class', axis=1)
y = dataset['Class']

Here the X variable contains all the columns from the dataset, except the "Class" column, which is the label. 

The y variable contains the values from the "Class" column. The X variable is our attribute set and y variable contains corresponding labels

In [10]:
X

Variance  Skewness  Curtosis  Entropy
0      3.62160   8.66610   -2.8073 -0.44699
1      4.54590   8.16740   -2.4586 -1.46210
2      3.86600  -2.63830    1.9242  0.10645
3      3.45660   9.52280   -4.0112 -3.59440
4      0.32924  -4.45520    4.5718 -0.98880
...        ...       ...       ...      ...
1367   0.40614   1.34920   -1.4501 -0.55949
1368  -1.38870  -4.87730    6.4774  0.34179
1369  -3.75030 -13.45860   17.5932 -2.77710
1370  -3.56370  -8.38270   12.3930 -1.28230
1371  -2.54190  -0.65804    2.6842  1.19520

[1372 rows x 4 columns]

In [11]:
y

0       0
1       0
2       0
3       0
4       0
       ..
1367    1
1368    1
1369    1
1370    1
1371    1
Name: Class, Length: 1372, dtype: int64

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

In [13]:
X_train.shape

(1097, 4)

In [15]:
y_train.shape

(1097,)

In [14]:
X_test.shape

(275, 4)

In [16]:
y_test.shape

(275,)

In [17]:
from sklearn.tree import DecisionTreeClassifier

classifier = DecisionTreeClassifier()
classifier.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [18]:
y_pred = classifier.predict(X_test)

**Evaluating the Algorithm**

In [20]:
from sklearn.metrics import classification_report, confusion_matrix

print(confusion_matrix(y_test, y_pred))

print()
print()
print(classification_report(y_test, y_pred))

[[163   1]
 [  1 110]]


              precision    recall  f1-score   support

           0       0.99      0.99      0.99       164
           1       0.99      0.99      0.99       111

    accuracy                           0.99       275
   macro avg       0.99      0.99      0.99       275
weighted avg       0.99      0.99      0.99       275



From the confusion matrix, you can see that out of 275 test instances, our algorithm misclassified only 2. This is 98.5 % accuracy. Not too bad!

#2. Decision Tree for Regression

The process of solving regression problem with decision tree using Scikit Learn is very similar to that of classification. 

However for regression we use DecisionTreeRegressor class of the tree library. 

Also the evaluation matrics for regression differ from those of classification. The rest of the process is almost same. 

In [21]:
#Importing Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [22]:
dataset = pd.read_csv('/content/petrol_consumption_new.csv')


In [23]:
dataset.head()

Petrol_tax  Average_income  ...  Population_Driver_licence(%)  Petrol_Consumption
0         9.0            3571  ...                         0.525                 541
1         9.0            4092  ...                         0.572                 524
2         9.0            3865  ...                         0.580                 561
3         7.5            4870  ...                         0.529                 414
4         8.0            4399  ...                         0.544                 410

[5 rows x 5 columns]

In [24]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Petrol_tax                    48 non-null     float64
 1   Average_income                48 non-null     int64  
 2   Paved_Highways                48 non-null     int64  
 3   Population_Driver_licence(%)  48 non-null     float64
 4   Petrol_Consumption            48 non-null     int64  
dtypes: float64(2), int64(3)
memory usage: 2.0 KB


In [25]:
dataset.describe()

Petrol_tax  ...  Petrol_Consumption
count   48.000000  ...           48.000000
mean     7.668333  ...          576.770833
std      0.950770  ...          111.885816
min      5.000000  ...          344.000000
25%      7.000000  ...          509.500000
50%      7.500000  ...          568.500000
75%      8.125000  ...          632.750000
max     10.000000  ...          968.000000

[8 rows x 5 columns]

In [26]:
# Preparing the Data

X = dataset.drop('Petrol_Consumption', axis=1)
y = dataset['Petrol_Consumption']

Here the X variable contains all the columns from the dataset, except 'Petrol_Consumption' column, which is the label. 

The y variable contains values from the 'Petrol_Consumption' column, which means that the X variable contains the attribute set and y variable contains the corresponding labels.

In [27]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

As mentioned earlier, for a regression task we'll use a different sklearn class than we did for the classification task. The class we'll be using here is the **DecisionTreeRegressor** class, as opposed to the **DecisionTreeClassifier** from before.

To train the tree, we'll instantiate the DecisionTreeRegressor class and call the fit method:

In [28]:
from sklearn.tree import DecisionTreeRegressor

regressor = DecisionTreeRegressor()
regressor.fit(X_train, y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=None, splitter='best')

In [29]:
y_pred = regressor.predict(X_test)

In [30]:
df=pd.DataFrame({'Actual':y_test, 'Predicted':y_pred})

df

Actual  Predicted
29     534      541.0
4      410      414.0
26     577      574.0
30     571      554.0
32     577      631.0
37     704      644.0
34     487      648.0
40     587      540.0
7      467      414.0
10     580      464.0

**Evaluating the Algorithm**

To evaluate performance of the regression algorithm, the commonly used metrics are **mean absolute error**, **mean squared error**, and **root mean squared error**. The Scikit-Learn library contains functions that can help calculate these values for us. To do so, use this code from the **metrics** package:

In [31]:
from sklearn import metrics

print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 52.2
Mean Squared Error: 5127.4
Root Mean Squared Error: 71.60586568152081


The mean absolute error for our algorithm is 52.2, which is less than 10 percent of the mean of all the values in the 'Petrol_Consumption' column. 

This means that our algorithm did a fine prediction job.